Rewrite the path close section to specify symmetric Abandon (#377)
quicwg/multipath: draft-ietf-quic-multipath.md

If validation succeeds, the client can continue to use the path.
If validation fails, the client MUST NOT use the path and can
remove any status associated to the path initiation attempt.
However, as the used Path ID is anyway consumed,

> and the endpoint MUST explicitly close the path as specified in

Reviewer: ```suggestion
the endpoint MUST explicitly close the path as specified in
```

a path, the peer can expect to not receive any more non-probing packets on
that path.

An endpoint that wants to close a path MUST explicitly
terminate the path by sending a PATH_ABANDON frame.
Note that while abandoning a path will cause
connection ID retirement, the inverse is not true: retiring the associated connection IDs
does not indicate path abandonment (see {{retire-cid-close}}).
This is true whether the decision to close the path results
from implicit signals such as idle time or packet losses
(see {(idle-time-close}}) or for any other reason, such as management
of local resources. It is also possible to abandon a path for which no
packet has been sent (see {{abandon-early}}).

When an endpoint receives a PATH_ABANDON frame, it MUST send a corresponding
PATH_ABANDON frame if it has not already done so, and it MUST treat all
connection identifiers received from the peer for that path as immediately
retired. However, the endpoint should retain some knowledge of the
connection identifiers sent to the peer for a short period,

> as specified in {{spurious-stateless-reset}}.

Author: I tried to develop the ACK_MP issue in the subsection "Handling ACK_MP for abandoned paths {#ack-after-abandon}". Maybe we should call that out here, in the same way that we call out stateless reset.

> as specified in {{spurious-stateless-reset}}.

Author: I also think that we have a hard rule on the data sender side: "if a path info is deleted, then all pending packets MUST be considered lost", and softer rule "don't delete the info too soon because you will end up with spurious retransmit". On the data receiver side, we could have a hard rule, "you MUST send an ACK_MP for all packets received before abandon by the peer", and a softer rule "you should wait a bit for packets to arrive after abandon so you can ACK them and avoid forcing the peer to resend them". Maybe develop that a bit more in the dedicated subsection.

> as specified in {{spurious-stateless-reset}}.

Author: @mirjak I think that the "tighten" commit fixes this. Please review.

a path, the peer can expect to not receive any more non-probing packets on
that path.

An endpoint that wants to close a path SHOULD explicitly
terminate the path by sending a PATH_ABANDON frame (see
{{path-abandon-close}}). Note that while abandoning a path will cause
An endpoint that wants to close a path MUST explicitly
terminate the path by sending a PATH_ABANDON frame.
Note that while abandoning a path will cause
connection ID retirement, the inverse is not true: retiring the associated connection IDs
does not indicate path abandonment (see further {{consume-retire-cid}}).
Implicit signals such as idle time or packet losses might be
the only way for an endpoint to detect path closure (see {(idle-time-close}})
if connectivity is broken on that path.

Note that other explicit closing mechanisms of {{QUIC-TRANSPORT}} still
apply on the whole connection. In particular, the reception of either a
CONNECTION_CLOSE ({{Section 10.2 of QUIC-TRANSPORT}}) or a Stateless

> Reset ({{Section 10.3 of QUIC-TRANSPORT}}) closes the connection.

Reviewer 2: @huitema why did you remove this paragraph?

Author: Can't remember why. Restore it if you want.

> does not indicate path abandonment (see {{retire-cid-close}}).

Reviewer 2: ```suggestion
does not indicate path abandonment (see further {{consume-retire-cid}}).
```

probing packet is received on a new path before sending the
CONNECTION_CLOSE frame.

### Avoiding Spurious Stateless Resets {#spurious-stateless-reset}

The peers that send a PATH_ABANDON frame MUST treat all connection
identifiers received from the peer for the path ID as immediately
retired. The Stateless Reset Tokens associated with these connection
identifiers MUST NOT be used to identify Stateless Reset packets
per {{Section 10.3 of QUIC-TRANSPORT}}.

Due to packet losses and network delays, packets sent on the path may
well arrive after the PATH_ABANDON frames have been sent or received.
If these packets arrive after the connection identifiers sent to the peer
have been retired, they will not be recognizes as bound for the local
connection and could trigger the peer to send a Stateless Reset
packet. The rule to "retain knowledge of connection ID for a short period"
is intended to reduce the risk of sending such spurious stateless
packets, but it cannot completely avoid that risk.

The immediate retirement of connection identifiers received for the
path guarantees that spurious stateless reset packets
sent by the peer will not cause the closure of the QUIC connection.

> ### Handling MP_ACK for abandoned paths {#ack-after-abandon}

Author: Sorry, updates lost in a github snafu. Resetting.

When an endpoint decides to send a PATH_ABANDON frame, there may
still be some unacknowledged packets. Some other packets may well
be in transit, and could be received shortly after sending the
PATH_ABANDON frame. As specified above, the endpoints SHOULD
send ACK_MP frames promptly, to avoid unnecessary data
retransmission after the peer deletes path resources.

> These ACK_MP frames SHOULD be send on a different path than the

Reviewer 2: I think this is a MUST. Or no normatively language here at all. I would rather say somewhere else that if the PATH_ABANDON frame was sent, the endpoint MUST NOT send on that path anymore.

Author: Yes, that's probably the best way to handle it, but there is the slight issue of abandoning the last path.

Reviewer 2: For the last path you should do connection close instead. That’s already written somewhere but it might be a SHOULD only. If so, we can maybe change to a MUST.

Author: Not sure it matters. Suppose that we treat "abandon of last path" as a protocol error. That means the peer receiving the abandon will automatically issue "connection close" with reason protocol error, and the peer sending the abandon will send nothing, because it must not send data after abandon. But yes, sending connection close is more graceful.

Author: Replace SHOULD by MUST.

probing packet is received on a new path before sending the
CONNECTION_CLOSE frame.

### Avoiding Spurious Stateless Resets {#spurious-stateless-reset}

The peers that send a PATH_ABANDON frame MUST treat all connection
identifiers received from the peer for the path ID as immediately
retired. The Stateless Reset Tokens associated with these connection
identifiers MUST NOT be used to identify Stateless Reset packets
per {{Section 10.3 of QUIC-TRANSPORT}}.

Due to packet losses and network delays, packets sent on the path may
well arrive after the PATH_ABANDON frames have been sent or received.
If these packets arrive after the connection identifiers sent to the peer

> have been retired, they will not be recognizes as bound for the local

Reviewer: ```suggestion
have been retired, they will not be recognized as bound for the local
```

periodic PING frames also helps prevent middlebox timeout, as discussed in
{{Section 10.1.2 of QUIC-TRANSPORT}}.

Server implementations need to select the sub-path idle timeout as a trade-
off between keeping resources, such as connection IDs, in use
for an excessive time or having to promptly re-establish a path
after a spurious estimate of path abandonment by the client.

Endpoints that desire to close a path because of the idle timer rule
MUST do so explicitly by sending a PATH_ABANDON frame, as defined in
{{path-close}}.

### Early Abandon {#abandon-early}

The are are scenarios in which an endpoint will receive a PATH_ABANDON frame
before receiving or sending any traffic on a path. For example, if the client
tries to initiate a path and the path cannot be establish, it will send a
PATH_ABANDON frame (see {{path-initiation}}). An endpoint may also decide

> to abandon a path for any reason, such as for example removing a hole from

Reviewer 3: ```suggestion
to abandon a path for any reason, for example, removing a hole from
```

periodic PING frames also helps prevent middlebox timeout, as discussed in
{{Section 10.1.2 of QUIC-TRANSPORT}}.

Server implementations need to select the sub-path idle timeout as a trade-
off between keeping resources, such as connection IDs, in use
for an excessive time or having to promptly re-establish a path
after a spurious estimate of path abandonment by the client.

Endpoints that desire to close a path because of the idle timer rule

> MUST do so explicitly by sending a PATH_ABANDON frame, as defined in

Reviewer 2: ```suggestion
MUST do so explicitly by sending a PATH_ABANDON frame on another active path, as defined in
```

Reviewer 3: I think it's better to allow endpoints to send PATH_ABANDON frame either on the same path or another active path. This depends on whether the same path is a blackhole. So we don't need to specify here.

on the path that the retired connection ID was used on but has
to use a different connection ID for the same Path ID when doing so.

> ## Packet Protection {#multipath-aead}

Reviewer 2: ```suggestion
# Packet Protection {#multipath-aead}
```

than 350ms, either the data packet or its ACK_MP were probably lost.

The simplest implementation is to compute smoothedRTT and RTTvar per

> {{Section 5.3 of QUIC-RECOVERY}} regardless of the path through which ACK_MPs are

Reviewer 2: ```suggestion
{{Section 5.3 of QUIC-RECOVERY}} regardless of the path through which ACK_MP frames are
```
